# Training resources

* [Babraham Bioinformatics](https://www.bioinformatics.babraham.ac.uk/training.html)

* [IMB Bioinformatics Core](https://bc.imb.sinica.edu.tw/BioinfoDocs/)

* [Harvard Chan Bioinformatics Core](https://bioinformatics.sph.harvard.edu/training)

* [Harvard FAS Informatics](https://informatics.fas.harvard.edu/category/tutorials.html)

* [UCDAVIS Bioinformatics Core](https://ucdavis-bioinformatics-training.github.io/)

* [UCI BioSci](https://swaruplab.bio.uci.edu/tutorial/tutorial.html)

* 

# Single cell data analysis

* [Single-cell best practices (python)](https://www.sc-best-practices.org/preamble.html)

* [Harvard FAS informatics nanocourse: scRNA-seq-workshop-Fall-2019](https://crazyhottommy.github.io/scRNA-seq-workshop-Fall-2019/)

* 

# Tools

* Seurat
https://satijalab.org/seurat/index.html

* Gene Set Enrichment Analysis (GSEA)

  * [singleseqgset: Single-cell RNAseq gene set enrichment analysis](https://arc85.github.io/singleseqgset/articles/singleseqgset.html)

  * fgsea: Fast Gene Set Enrichment Analysis
    * [ctlat](https://github.com/ctlab/fgsea)

  * [GSEA annotation](https://rpubs.com/eraz0001/GSEA_annotation) 

* VDJ analysis
  * [scRepertoire](https://ncborcherding.github.io/vignettes/vignette.html)









