# Sequence file formats : Fasta and fastq.

* Fasta format

    * header : >, a number of rows
    * used for bothe nucleic acids and proteins
    
![image](https://user-images.githubusercontent.com/49796592/129100577-93625ae0-ffdf-4a9e-8072-1dd0d3b505a9.png)

* FASTQ format (files with an extension .fastq)
    * Header : containing @ as the first character
    * Sequence : the sequence content
    * a spacer
    * Quality : encoded using ASCII characters

![image](https://user-images.githubusercontent.com/49796592/129100934-5577137f-f4db-4fc9-91bf-de3ac2f920d1.png)

* Phred quality score

![image](https://user-images.githubusercontent.com/49796592/129101445-d7985cd1-2755-48c0-85ff-1b546b6d1003.png)

    * Score = 10 (symbol ‘+’) => probability of incorrect base call = 0.1 => base call accuracy = 90%
    * Score = 20 (symbol ‘5’) => probability of incorrect base call = 0.01 => base call accuracy = 99%
    * Score = 30 (symbol ‘?’) => probability of incorrect base call = 0.001 => base call accuracy = 99.9% - This is a commonly acceptable threshold for trimming.
    * Score = 40 (symbol ‘I’) => probability of incorrect base call = 0.0001 => base call accuracy = 99.99%


# 2 FastQC

* FastQC calculates statistics about the composition and quality of raw sequences.

* Run FastQC in an Interactively mode

    * Instrall the FastQC application on MacOSX
    * Open the FastQC and load FastQ files
    * View the QC results and save the report

* Run FastQC in a non-interactively mode (as part of a pipeline)

    * Instrall the FastQC by downloading the file
    * run FastQC in terminal
    * Generate automatically report file in the same folder
    
```
$ fastqc A.read1.fq
Approx 5% complete for A.read1.fq
Approx 10% complete for A.read1.fq
Approx 15% complete for A.read1.fq
Approx 20% complete for A.read1.fq
Approx 25% complete for A.read1.fq
Approx 30% complete for A.read1.fq
Approx 35% complete for A.read1.fq
Approx 40% complete for A.read1.fq
Approx 45% complete for A.read1.fq
Approx 50% complete for A.read1.fq
Approx 55% complete for A.read1.fq
Approx 60% complete for A.read1.fq
Approx 65% complete for A.read1.fq
Approx 70% complete for A.read1.fq
Approx 75% complete for A.read1.fq
Approx 80% complete for A.read1.fq
Approx 85% complete for A.read1.fq
Approx 90% complete for A.read1.fq
Approx 95% complete for A.read1.fq
Analysis complete for A.read1.fq
```

* Display the results with a browser 

```
$ open -a "Google Chrome" A.read1_fastqc.html
```

# 3 Fastq Screen

* Fastq Screen looks for possible contaminations.

* Download genome files

```
$ fastq_screen --get_genomes
```

# 4 Initial processing of sequencing reads

* Demultiplex by index or barcode (it is usually done in the sequencing facility)
* Remove adapter sequences
* Trim reads by quality
* Discard reads by quality/ambiguity
* Filter reads by k-mer coverage (recommended for the de novo assembly)
* Normalize k-mer coverage (recommended for the de novo assembly)










