# 1 Genome/Transcriptome reference

## 1.1 Get the reference genome or transcriptome from a public database

* A FASTA file of the genome/transcriptome sequence (a file with an extension .fa)

* A GIF/GFF file with annotation (a file with an extension .gtf or .gff)

## 1.2 Public resources on genome/transcriptome sequences and annotations

* Gencode  https://www.gencodegenes.org/
* Ensembl  https://useast.ensembl.org/index.html
* UCSC Genome Browser  https://genome.ucsc.edu/ 

## 1.3 FASTA and GTF/GFF data formats

* The genome is generally represented as a FASTA file (.fa file) with the header indicated by the “>”

* The annotation is stored in General Transfer Format (GTF) format (which is an extension of the older GFF format)

# 2 Tools for read mapping

## 2.1 Fast (splice-unaware) aligners to a reference transcriptome

* align short reads to a transcriptome reference

### Bowtie

### Bowtie2

### BWA

### GEM

## 2.2 Splice-aware aligners to a reference genome

### Tophat

### HISAT2

### STAR(Spliced Transcripts Alignment to a Reference)

* Create a genome index
    * Download the reference files : the genome sequence in FASTA format and the annotation in GTF format
    * unzip the reference files using  
    * The basic options for aligning reads to the genome using STAR are:

        * `--runThreadN`: number of threads / cores
        * `--readFilesIn`: /path/to/FASTQ_file
        * `--genomeDir`: /path/to/genome_indices_directory
        * `--outFileNamePrefix`: prefix for all output files

* Map reads to the genome




## 2.3 Quasi-mappers (alignment-free mappers) to a reference transcriptome

### Sailfish

### Salmon

### Kallisto

