# 1 Genome/Transcriptome reference

## 1.1 Get the reference genome or transcriptome from a public database

* A FASTA file of the genome/transcriptome sequence (a file with an extension .fa)

* A GIF/GFF file with annotation (a file with an extension .gtf or .gff)

## 1.2 Public resources on genome/transcriptome sequences and annotations

* Gencode  https://www.gencodegenes.org/
* Ensembl  https://useast.ensembl.org/index.html
* UCSC Genome Browser  https://genome.ucsc.edu/ 

## 1.3 FASTA and GTF/GFF data formats

* The genome is generally represented as a FASTA file (.fa file) with the header indicated by the “>”

* The annotation is stored in General Transfer Format (GTF) format (which is an extension of the older GFF format)

# 2 Tools for read mapping

## 2.1 Fast (splice-unaware) aligners to a reference transcriptome

* align short reads to a transcriptome reference

### Bowtie

### Bowtie2

### BWA

### GEM

## 2.2 Splice-aware aligners to a reference genome

### Tophat

> a fast splice junction mapper for RNA-Seq reads. It aligns RNA-Seq reads to mammalian-sized genomes using the ultra high-throughput short read aligner Bowtie, and then analyzes the mapping results to identify splice junctions between exons.

* Prepare your reference

    * Usage: `bowtie2-build [options]* <reference_in> <bt2_base>`
    
```
bowtie2-build --large-index GRCm39.genome.fa mouse_genome_GRCm39.btindex
```
* Prepare your reads

* Run Tophat

```
tophat -p4 -G 04_Mapping/reference/gencode.vM27.annotation.gtf 04_Mapping/reference/GRCm39/mouse_genome_GRCm39.btindex 03_Trimming/skewer/A.read1.fq-trimmed.fastq -o 04_Mapping/alignments/tophat
```

### HISAT2

* the next generation of spliced aligner from the same group that have developed TopHat. It is a fast and sensitive alignment program for mapping next-generation sequencing reads (both DNA and RNA) to a population of human genomes (as well as to a single reference genome). The indexing scheme is called a Hierarchical Graph FM index (HGFM).

* Install Hisat2

```
brew install hisat2
```

* Create Hisat2 indexes

```
hisat2-build -p 4 reference/GRCm39.genome.fa alignments/hisat2/GRCm39.genome.index
```

* Map the reads for each sample to the reference genome

```
hisat2 -p 4 --dta -x 04_Mapping/alignments/hisat2/Index/GRCm39.genome.index 03_Trimming/skewer/F.read1.fq-trimmed.fastq -S 04_Mapping/alignments/hisat2/F_skewer_hisat2.sam
```

* Sort and convert the SAM files to BAM

```
samtools sort -@ 4 -o F_skewer_hisat2.bam F_skewer_hisat2.sam
```

* Assemble and quantify expressed genes and transcripts

```





### STAR(Spliced Transcripts Alignment to a Reference)

> an ultrafast universal RNA-seq aligner. It uses sequential maximum mappable seed search in uncompressed suffix arrays followed by seed clustering and stitching procedure. It is also able to search for gene fusions.

* Create a genome index
    * Download the reference files : the genome sequence in FASTA format and the annotation in GTF format
    * unzip the reference files using  
    * The basic options for aligning reads to the genome using STAR are:

        * `--runThreadN`: number of threads / cores
        * `--readFilesIn`: /path/to/FASTQ_file
        * `--genomeDir`: /path/to/genome_indices_directory
        * `--outFileNamePrefix`: prefix for all output files

```
STAR --runMode genomeGenerate --genomeDir 04_Mapping/indexes/genome --genomeFastaFiles 04_Mapping/reference/GRCm39.primary_assembly.genome.fa --sjdbGTFfile 04_Mapping/reference/gencode.vM27.annotation.gtf --sjdbOverhang 50 --outFileNamePrefix 04_Mapping/alignments/star --runThreadN 4
```

* Map reads to the genome




## 2.3 Quasi-mappers (alignment-free mappers) to a reference transcriptome

### Sailfish

### Salmon : a tool for quantifying the expression of transcripts using RNA-seq data

> an advanced version of Sailfish, by the same authors, tool for wicked-fast transcript quantification from RNA-seq data. It requires a set of target transcripts to quantify and a K-mer parameter to make the index (i.e. minimum acceptable alignment).

* Building the Salmon index

    * Get the transcript sequences in the FASTA format
   
```
$ salmon index -t gencode.vM27.transcripts.fa.gz -i transcripts_index
```

* Quantifying the samples

```
salmon quant -i indexes/transcripts -l SR  -r ../03*/fastp/D*.gz -o alignments/salmon/D -g reference/gencode.vM27.annotation.gtf --seqBias --validateMappings -p 4 
```


### Kallisto







# 3 BAM/SAM/CRAM format

* BAM format : a compressed version of the SAM format
* conver BAM to SAM : using samtools

# 4 Alignment QC 

## 4.1 QualiMap

* usage : `usage: qualimap rnaseq [-a <arg>] -bam <arg> -gtf <arg> [-oc <arg>] [-outdir
      <arg>] [-outfile <arg>] [-outformat <arg>] [-p <arg>]`

* IMPORTANT: before running QualiMap ensure enough disk space for a temporary directory ./tmp that the program is required, running the following command:

`export _JAVA_OPTIONS="-Djava.io.tmpdir=./tmp -Xmx6G"`

* Run QualiMap rnaseq

```
qualimap rnaseq -bam alignments/tophat/A/accepted_hits.bam -gtf reference/gencode.vM27.annotation.gtf -outdir qc_alignments 
```

# 5 Genome Broweser

## 5.1 Web-based genome browsers

* UCSC Genome Broswer
* Ensembl Genome Browser
* NCBI Genome Data Viewer

## 5.2 Desktop applications

* JBrowse
* GBrowse
* IGV



# 6 Read Counting

## 6.1 Counting with Cufflinks

* Cufflinks is part of the Bowtie/Tophat suite - is one tool thatcan be used for assembling and quantifying transcripts from reads aligned to a reference genome. While it is possible to provide an annnotation file to guide the quantification, we will not be doing so now.

* Install cufflinks by conda

```
conda create --name py36 python=3.6
conda activate py36
conda install cufflinks
```

* Run cufflinks : assemble transcripts for each sample

```R
% cufflinks -p 4 --library-type fr-unstranded -o ./04_Mapping/alignments/tophat/E/cufflinks -G ./04_Mapping/reference/gencode.vM27.annotation.gtf ./04_Mapping/alignments/tophat/E/accepted_hits.bam
```

```R
# if error occured, try to convert bam to sam
% samtools view -h 04_Mapping/alignments/tophat/Z/accepted_hits.bam > 04_Mapping/alignments/tophat/Z/Z_tophat.sam 

% cufflinks -p 4 --library-type fr-unstranded -o ./04_Mapping/alignments/tophat/Z/cufflinks -G ./04_Mapping/reference/gencode.vM27.annotation.gtf ./04_Mapping/alignments/tophat/Z/Z_tophat.sam
```

* Create a file called assemblies.txt that lists the assembly file for each sample.

* Run Cuffmerge

```
% cuffmerge -g *.gtf -s *.fa -p 4 assemblies.txt
```






## 6.2 Counting with Rsubread

As you might imagine, running Cufflinks for all our samples to assemble transcripts would take a while, In our case,
however, we are only interested in looking for differential expression in genes that we have already defined. Believe it or
not, we can do this in Bioconductor – so let’s move back to our trusty R! We will use the package Rsubread[5] to count
the reads mapping to genes in the human genome.

```R
library(Rsubread)
filesToCount <- dir("bam", pattern=".bam$", full.names=T)
```
Rsubread has a number of inbuild annotations that we can make use of. Once again, there are a number of options to
play around with, although we will keep things simple for now and generate a count matrix to use in the practical on
differential expression with edgeR.

```R
tmp <- featureCounts(filesToCount, annot.inbuilt="hg19", ignoreDup=F)
save(tmp, file="../Day3/countMatrix.RData")
```








# 6.3 Counting with featureCounts



# 6.4 Counting with htseq-count








